source/adapters/level_zero/v2/command_list_manager.cpp

```diff
@@ -0,0 +1,146 @@
+  TRACK_SCOPE_LATENCY("ur_command_list_manager::appendKernelLaunch");
+
+  UR_ASSERT(hKernel, UR_RESULT_ERROR_INVALID_NULL_HANDLE);
+  UR_ASSERT(hKernel->getProgramHandle(), UR_RESULT_ERROR_INVALID_NULL_POINTER);
+
+  UR_ASSERT(workDim > 0, UR_RESULT_ERROR_INVALID_WORK_DIMENSION);
+  UR_ASSERT(workDim < 4, UR_RESULT_ERROR_INVALID_WORK_DIMENSION);
+
+  ze_kernel_handle_t hZeKernel = hKernel->getZeHandle(device);
+
+  std::scoped_lock<ur_shared_mutex, ur_shared_mutex> Lock(this->Mutex,
+                                                          hKernel->Mutex);
+
+  ze_group_count_t zeThreadGroupDimensions{1, 1, 1};
+  uint32_t WG[3]{};
+  UR_CALL(calculateKernelWorkDimensions(hZeKernel, device,
+                                        zeThreadGroupDimensions, WG, workDim,
+                                        pGlobalWorkSize, pLocalWorkSize));
+
+  auto signalEvent = getSignalEvent(phEvent, UR_COMMAND_KERNEL_LAUNCH);
+
+  auto waitList = getWaitListView(phEventWaitList, numEventsInWaitList);
+
+  bool memoryMigrated = false;
+  auto memoryMigrate = [&](void *src, void *dst, size_t size) {
+    ZE2UR_CALL_THROWS(zeCommandListAppendMemoryCopy,
+                      (zeCommandList.get(), dst, src, size, nullptr,
+                       waitList.second, waitList.first));
+    memoryMigrated = true;
+  };
+
+  UR_CALL(hKernel->prepareForSubmission(context, device, pGlobalWorkOffset,
+                                        workDim, WG[0], WG[1], WG[2],
+                                        memoryMigrate));
+
+  if (memoryMigrated) {
+    // If memory was migrated, we don't need to pass the wait list to
+    // the copy command again.
+    waitList.first = nullptr;
+    waitList.second = 0;
+  }
+
+  TRACK_SCOPE_LATENCY(
+      "ur_command_list_manager::zeCommandListAppendLaunchKernel");
+  auto zeSignalEvent = signalEvent ? signalEvent->getZeEvent() : nullptr;
+  ZE2UR_CALL(zeCommandListAppendLaunchKernel,
+             (zeCommandList.get(), hZeKernel, &zeThreadGroupDimensions,
+              zeSignalEvent, waitList.second, waitList.first));
+
+  return UR_RESULT_SUCCESS;
+}
+ur_result_t ur_command_list_manager::enqueueGenericCommandListsExp(
+    uint32_t numCommandLists, ze_command_list_handle_t *phCommandLists,
+    ur_event_handle_t *phEvent, uint32_t numEventsInWaitList,
+    const ur_event_handle_t *phEventWaitList, ur_command_t callerCommand) {
+
+  std::scoped_lock<ur_shared_mutex> Lock(this->Mutex);
+  auto signalEvent = getSignalEvent(phEvent, callerCommand);
+
+  auto [pWaitEvents, numWaitEvents] =
+      getWaitListView(phEventWaitList, numEventsInWaitList);
+
+  auto zeSignalEvent = signalEvent ? signalEvent->getZeEvent() : nullptr;
+
+  ZE2UR_CALL(zeCommandListImmediateAppendCommandListsExp,
+             (zeCommandList.get(), numCommandLists, phCommandLists,
+              zeSignalEvent, numWaitEvents, pWaitEvents));
+
+  return UR_RESULT_SUCCESS;
+}
+
+ur_result_t ur_command_list_manager::closeCommandList() {
```

Comment:
this should be inside of command buffer. Immediate command lists never need to be closed.